layabout.py

```diff
@@ -1,3 +1,287 @@
+        return decorator
+
+    def connect(self, token: str = None) -> bool:
+        """
+        Attempt to establish or reset a connection to Slack's API.
+
+        Note:
+            It isn't normally necessary to use this method as
+            :obj:`Layabout.run` will take care of establishing a connection for
+            you. This is mostly helpful if you want to take advantage of the
+            embedded :obj:`slackclient.SlackClient` on the :obj:`Layabout`
+            instance *before* entering into the :obj:`Layabout.run` loop.
+
+        Args:
+            token: A Slack API token. If given it will override an existing
+                connection (if one exists), otherwise an existing token or
+                an environment variable will be used.
+
+        Returns:
+            Whether the connection succeeded.
+
+        Raises:
+            MissingSlackToken: If no API token is available.
+        """
+        resetting = False
+
+        # If we were given a token let's prefer the new one and establish or
+        # reset the connection.
+        if token is not None:
+            self._token = token
+            resetting = True
+
+        # We don't have an existing token, so let's try to get one from an
+        # environment variable or give up.
+        if self._token is None:
+            try:
+                self._token = os.environ[self._env_var]
+            except KeyError:
+                raise MissingSlackToken('Cannot connect to the Slack API'
+                                        ' without a token')
+
+        # Either we've never connected before or we're purposefully resetting
+        # the connection.
+        if self.slack is None or resetting:
+            # TODO: Maybe set an appropriate user agent string here using
+            # self.name.
+            self.slack = SlackClient(token=self._token)
+
+        # Use whatever token we've got to attempt to connect (or reconnect).
+        return self.slack.rtm_connect()
+
+    def _reconnect(self, retries: int,
+                   backoff: Callable[[int], float]) -> bool:
+        """
+        Attempt to reconnect to the Slack API.
+
+        Args:
+            retries: The number of retry attempts to make if a connection
+                to Slack if not established or is lost.
+            backoff: The strategy used to determine how
+                long to wait between retries. Must take as input the number of
+                the current retry and output a :obj:`float`.
+
+        Returns:
+            Whether the reconnection succeeded.
+        """
+        # TODO: Should retries start at 0 or 1?
+        for retry in range(retries):
+            if self.connect():
+                return True
+            else:
+                interval = backoff(retry)
+                log.debug("Waiting %.3fs before retrying", interval)
+                time.sleep(interval)
+
+        return False
+
+    def run(self, token: str = None, interval: float = 0.5,
```

Comment:
In Python `getToken` will only be called once; `getToken` would only be called at function definition time. Therefor it could be a constant or an attribute but a function is never the right call.

Reply:
I don't believe it's true that `getToken` would only be called at function definition time. I think abstracting token acquisition might actually be useful here.

Reply:
No, @rawrgulmuffins is right. You can't put a call to a function in the default argument position.  You'd have to leave it as `None`, and do the same thing I suggested earlier with a ` self._token = token or self.get_token()`

> Default parameter values are evaluated from left to right when the function definition is executed. This means that the expression is evaluated once, when the function is defined, and that the same “pre-computed” value is used for each call.

https://docs.python.org/3/reference/compound_stmts.html#function-definitions

Reply:
Oh, I totally misunderstood why @kyle-rader was suggesting then. I didn't think he meant for the call to be in the signature. Yeah, @rawrgulmuffins is absolutely right. :+1: